File: src/main/java/com/atsuishio/superbwarfare/init/ModCapabilities.java
package com.atsuishio.superbwarfare.init;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.block.entity.ChargingStationBlockEntity;
import com.atsuishio.superbwarfare.block.entity.CreativeChargingStationBlockEntity;
import com.atsuishio.superbwarfare.block.entity.FuMO25BlockEntity;
import com.atsuishio.superbwarfare.capability.api.IEnergyStorage;
import com.atsuishio.superbwarfare.capability.api.IItemHandler;
import com.atsuishio.superbwarfare.capability.api.InvWrapper;
import com.atsuishio.superbwarfare.capability.api.SidedInvWrapper;
import com.atsuishio.superbwarfare.capability.energy.ItemEnergyStorage;
import com.atsuishio.superbwarfare.capability.laser.LaserCapability;
import com.atsuishio.superbwarfare.capability.laser.LaserCapabilityProvider;
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity;
import com.atsuishio.superbwarfare.item.CreativeChargingStationBlockItem;
import com.atsuishio.superbwarfare.item.EnergyStorageItem;
import net.fabricmc.fabric.api.lookup.v1.block.BlockApiLookup;
import net.fabricmc.fabric.api.lookup.v1.entity.EntityApiLookup;
import net.fabricmc.fabric.api.lookup.v1.item.ItemApiLookup;
import net.minecraft.core.Direction;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.world.entity.EntityType;

public class ModCapabilities {

    public static final EntityApiLookup<LaserCapability, Void> LASER_CAPABILITY = EntityApiLookup.get(Mod.loc("laser_capability"), LaserCapability.class, Void.class);

    public static final BlockApiLookup<IEnergyStorage, Direction> ENERGY_BLOCK = BlockApiLookup.get(Mod.loc("energy"), IEnergyStorage.class, Direction.class);

    public static final ItemApiLookup<IEnergyStorage, Void> ENERGY_ITEM = ItemApiLookup.get(Mod.loc("energy"), IEnergyStorage.class, Void.class);

    public static final EntityApiLookup<IEnergyStorage, Void> ENERGY_ENTITY = EntityApiLookup.get(Mod.loc("energy"), IEnergyStorage.class, Void.class);

    public static final BlockApiLookup<IItemHandler, Direction> ITEM_HANDLER_BLOCK = BlockApiLookup.get(Mod.loc("item_handler"), IItemHandler.class, Direction.class);

    public static final EntityApiLookup<IItemHandler, Void> ITEM_HANDLER_ENTITY = EntityApiLookup.get(Mod.loc("item_handler"), IItemHandler.class, Void.class);

    public static void init() {
        LASER_CAPABILITY.registerForTypes(new LaserCapabilityProvider(), EntityType.PLAYER);

        ENERGY_BLOCK.registerForBlockEntity(ChargingStationBlockEntity::getEnergyStorage, ModBlockEntities.CHARGING_STATION.value());
        ENERGY_BLOCK.registerForBlockEntity(CreativeChargingStationBlockEntity::getEnergyStorage, ModBlockEntities.CREATIVE_CHARGING_STATION.value());
        ENERGY_BLOCK.registerForBlockEntity(FuMO25BlockEntity::getEnergyStorage, ModBlockEntities.FUMO_25.value());

        ENERGY_ITEM.registerForItems((stack, ctx) -> ((CreativeChargingStationBlockItem) stack.getItem()).getEnergyStorage(), ModItems.CREATIVE_CHARGING_STATION.value());

        ITEM_HANDLER_BLOCK.registerForBlockEntity((ChargingStationBlockEntity be, Direction ctx) -> {
            if (ctx == null || be.isRemoved()) return null;

            var itemHandlers = new IItemHandler[]{
                    new SidedInvWrapper(be, Direction.UP),
                    new SidedInvWrapper(be, Direction.DOWN),
                    new SidedInvWrapper(be, Direction.NORTH),
            };

            return switch (ctx) {
                case UP -> itemHandlers[0];
                case DOWN -> itemHandlers[1];
                default -> itemHandlers[2];
            };
        }, ModBlockEntities.CHARGING_STATION.value());

        ITEM_HANDLER_BLOCK.registerForBlockEntities((object, context) -> new InvWrapper(object), ModBlockEntities.SUPERB_ITEM_INTERFACE.get());

        for (var item : BuiltInRegistries.ITEM) {
            if (item instanceof EnergyStorageItem energyItem) {
                ENERGY_ITEM.registerForItems((stack, ctx) -> new ItemEnergyStorage(
                        stack,
                        s -> energyItem.getMaxEnergy(s),
                        s -> energyItem.getMaxReceiveEnergy(s),
                        s -> energyItem.getMaxExtractEnergy(s)
                ), item);
            }
        }

        for (var entity : BuiltInRegistries.ENTITY_TYPE) {
            ENERGY_ENTITY.registerForTypes((obj, ctx) ->
                    (obj instanceof VehicleEntity vehicle && vehicle.hasEnergyStorage()) ? vehicle.getEnergyStorage() : null, entity);

            ITEM_HANDLER_ENTITY.registerForTypes((obj, ctx) ->
                    (obj instanceof VehicleEntity vehicle && vehicle.hasContainer()) ? new InvWrapper(vehicle) : null, entity);
        }

        ENERGY_ENTITY.registerForTypes((obj, ctx) -> obj.getEnergyStorage(), ModEntities.DPS_GENERATOR.get());
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/init/ModCapabilities.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.block.entity.ChargingStationBlockEntity
import com.atsuishio.superbwarfare.block.entity.BlueprintResearchTableBlockEntity
import com.atsuishio.superbwarfare.block.entity.CreativeChargingStationBlockEntity
import com.atsuishio.superbwarfare.block.entity.FuMO25BlockEntity
import com.atsuishio.superbwarfare.block.entity.SuperbItemInterfaceBlockEntity
import com.atsuishio.superbwarfare.capability.api.IItemHandler
import com.atsuishio.superbwarfare.capability.api.InvWrapper
import com.atsuishio.superbwarfare.capability.api.SidedInvWrapper
import com.atsuishio.superbwarfare.capability.energy.ItemEnergyStorage
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity
import com.atsuishio.superbwarfare.entity.living.DPSGeneratorEntity
import com.atsuishio.superbwarfare.item.EnergyStorageItem
import com.atsuishio.superbwarfare.item.blockitem.CreativeChargingStationBlockItem
import net.fabricmc.fabric.api.lookup.v1.block.BlockApiLookup
import net.fabricmc.fabric.api.lookup.v1.entity.EntityApiLookup
import net.minecraft.core.Direction
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.entity.player.Player
import team.reborn.energy.api.EnergyStorage
import team.reborn.energy.api.base.SimpleEnergyItem

object ModCapabilities {
    @JvmField
    val ENERGY_BLOCK = EnergyStorage.SIDED

    @JvmField
    val ENERGY_ITEM = EnergyStorage.ITEM

    @JvmField
    val ITEM_HANDLER_BLOCK: BlockApiLookup<IItemHandler, Direction?> =
        BlockApiLookup.get(Mod.loc("item_handler"), IItemHandler::class.java, Direction::class.java)

    @JvmField
    val ITEM_HANDLER_ENTITY: EntityApiLookup<IItemHandler, Void?> =
        EntityApiLookup.get(Mod.loc("item_handler"), IItemHandler::class.java, Void::class.java)

    @JvmStatic
    fun init() {
        EnergyStorage.SIDED.registerForBlockEntity(
            ChargingStationBlockEntity::getEnergyStorage,
            ModBlockEntities.CHARGING_STATION
        )

        EnergyStorage.SIDED.registerForBlockEntity(
            CreativeChargingStationBlockEntity::getEnergyStorage,
            ModBlockEntities.CREATIVE_CHARGING_STATION
        )

        EnergyStorage.SIDED.registerForBlockEntity(
            { be: FuMO25BlockEntity, _: Direction? -> be.getEnergyStorage() },
            ModBlockEntities.FUMO_25
        )

        EnergyStorage.ITEM.registerForItems(
            { stack, _ -> (stack.item as CreativeChargingStationBlockItem).energyStorage },
            ModItems.CREATIVE_CHARGING_STATION
        )

        ITEM_HANDLER_BLOCK.registerForBlockEntity(
            { be: ChargingStationBlockEntity, ctx: Direction? ->
                if (ctx == null || be.isRemoved) return@registerForBlockEntity null

                val itemHandlers = arrayOf<IItemHandler>(
                    SidedInvWrapper(be, Direction.UP),
                    SidedInvWrapper(be, Direction.DOWN),
                    SidedInvWrapper(be, Direction.NORTH)
                )

                when (ctx) {
                    Direction.UP -> itemHandlers[0]
                    Direction.DOWN -> itemHandlers[1]
                    else -> itemHandlers[2]
                }
            },
            ModBlockEntities.CHARGING_STATION
        )

        ITEM_HANDLER_BLOCK.registerForBlockEntity(
            { obj: SuperbItemInterfaceBlockEntity, _ ->
                InvWrapper(obj)
            },
            ModBlockEntities.SUPERB_ITEM_INTERFACE
        )

        ITEM_HANDLER_BLOCK.registerForBlockEntity(
            { be: BlueprintResearchTableBlockEntity, ctx: Direction? ->
                if (be.isRemoved) null else SidedInvWrapper(
                    be,
                    when (ctx) {
                        Direction.UP, Direction.DOWN, Direction.NORTH, Direction.SOUTH -> ctx
                        else -> Direction.EAST
                    }
                )
            },
            ModBlockEntities.BLUEPRINT_RESEARCH_TABLE
        )

        for (item in BuiltInRegistries.ITEM) {
            if (item is EnergyStorageItem) {
                EnergyStorage.ITEM.registerForItems(
                    { stack, context ->
                        if (context == null) {
                            ItemEnergyStorage(
                                stack,
                                item::getMaxEnergy,
                                item::getMaxReceiveEnergy,
                                item::getMaxExtractEnergy
                            )
                        } else {
                            SimpleEnergyItem.createStorage(
                                context,
                                item.getMaxEnergy(stack).toLong(),
                                item.getMaxReceiveEnergy(stack).toLong(),
                                item.getMaxExtractEnergy(stack).toLong()
                            )
                        }
                    },
                    item
                )
            }
        }

        for (entity in BuiltInRegistries.ENTITY_TYPE) {
            ITEM_HANDLER_ENTITY.registerForType(
                { obj, _ ->
                    when {
                        obj is Player -> InvWrapper(obj.inventory)
                        obj is VehicleEntity && obj.hasContainer() -> obj.inventory
                        else -> null
                    }
                },
                entity
            )
        }
    }

    @JvmStatic
    fun getEntityEnergyStorage(entity: net.minecraft.world.entity.Entity): EnergyStorage? = when {
        entity is DPSGeneratorEntity -> entity.energyStorage
        entity is VehicleEntity && entity.hasEnergyStorage() -> entity.getEnergyStorage()
        else -> null
    }
}
